cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -1134,6 +1134,37 @@ ArrayKernelExec GeneratePhysicalInteger(detail::GetTypeId get_id) {
   }
 }
 
+template <template <typename... Args> class KernelGenerator, typename Op>
+ArrayKernelExec ArithmeticExecFromOp(detail::GetTypeId get_id) {
```

Comment:
Need to add a template parameter (a parameter pack) to allow `CumulativeSumOptions` to be passed to the `KernelGenerator`. For example,
```c++
template <template <typename...> class KernelGenerator, typename Op, typename... Args>
ArrayKernelExec ArithmeticExecFromOp(detail::GetTypeId get_id) {
...
  case Type::INT8:
      return KernelGenerator<Int8Type, Int8Type, Op, Args...>::Exec;
...
}
```